binWlz/WlzHistogramMatchObj.1
'\" t
.\" ident MRC HGU $Id$
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" Project:    Woolz
.\" Title:      WlzHistogramMatchObj.1
.\" Date:       March 1999
.\" Author:     Bill Hill
.\" Copyright:	1999 Medical Research Council, UK.
.\"		All rights reserved.
.\" Address:	MRC Human Genetics Unit,
.\"		Western General Hospital,
.\"		Edinburgh, EH4 2XU, UK.
.\" Purpose:    Woolz binary which histogram matches a Woolz domain
.\"		object.
.\" $Revision$
.\" Maintenance:Log changes below, with most recent at top of list.
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.TH "WlzHistogramMatchObj" 1 "MRC HGU Woolz" "Woolz Procedure Library"
.SH NAME
WlzHistogramMatchObj \- Histogram matches a Woolz domain object.
.SH SYNOPSIS
.LP
.BI "WlzHistogramMatchObj  -m" "<target> "
.in +4m
.br
.BI "[ -o" "<output object file>" "] [-h]"
.br
.B "[-D]"
.BI "[-s" "<smoothing>" "]"
.BI "[-t" "<smoothing>" "]"
.br
.BI "[-i] [-d" "<min>,<max>" "]"
.BI [ "<input object file>" ]
.in -4m
.SH OPTIONS
.TS
tab(^);
lb l.
\-D^dither pixel values when mapping.
\-m^target histogram object file.
\-o^output object file.
\-h^help, prints usage message.
\-s^input object histogram smoothing.
\-t^target histogram smoothing.
\-i^independent planes.
\-d^histogram distance range.
.TE
By default the input object is read from the standard input
and the output object  is written to the standard output.
.LP
The histogram distance range limits
are comma separated and default to 0.0,1.0.
.LP
If a histogram distance range is given,
then only the domain objects/planes
which have a histogram within 
that distance range from the target histogram
are matched.
.SH DESCRIPTION
Modifies the grey values of the input object
so that the histogram of the modified object's grey values
matches the target histogram.
.LP
The smoothing factor is the low\-pass gaussian convolution
kernel half\-height full\-width in bins
(default 0), used to smooth the histogram(s)
before matching.
.SH EXAMPLES
.LP
.ps -2
.cs R 24
.nf

# An example which uses WlzHistogramMatchObj and WlzHistogramObj to
# match the histogram of one domain object to another.
# an object.

WlzHistogramObj refObj.wlz | \\
WlzHistogramMatchObj -m - -o matchedObj.wlz myObj.wlz

# An example which uses WlzHistogramMatchObj to match all planes of
# a 3D image to plane 100.

WlzClipObjToBox -z 100,100 my3Dobj.wlz| \\
WlzHistogramObj - | \\
WlzHistogramMatchObj -m - -o matched3DObj.wlz my3Dobj.wlz

.fi
.cs R
.ps +2
.SH SEE ALSO
WlzHistogramObj(1),
WlzHistogramDistance(1),
WlzHistogram(3).
.SH BUGS
Still to be found.


ChangeLog
$Log$
Revision 1.2  2007/12/05 13:33:30  bill
*** empty log message ***



binWlz/WlzHistogramMatchObj.c
#if defined(__GNUC__)
#ident "MRC HGU $Id$"
#else
#if defined(__SUNPRO_C) || defined(__SUNPRO_CC)
#pragma ident "MRC HGU $Id$"
#else
static char _WlzHistogramMatchObj_c[] = "MRC HGU $Id$";
#endif
#endif
/*!
* \file         binWlz/WlzHistogramMatchObj.c
* \author       Bill Hill
* \date         March 1999
* \version      $Id$
* \par
* Address:
*               MRC Human Genetics Unit,
*               Western General Hospital,
*               Edinburgh, EH4 2XU, UK.
* \par
* Copyright (C) 2005 Medical research Council, UK.
* 
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be
* useful but WITHOUT ANY WARRANTY; without even the implied
* warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
* PURPOSE.  See the GNU General Public License for more
* details.
*
* You should have received a copy of the GNU General Public
* License along with this program; if not, write to the Free
* Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
* Boston, MA  02110-1301, USA.
* \brief	Histogram matches a domain object.
* \ingroup	BinWlz
* \todo         -
* \bug          None known.
*
* \par Binary
* \ref wlzhistogrammatchobj "WlzHistogramMatchObj"
*/

/*!
\ingroup      BinWlz
\defgroup     wlzhistogrammatchobj WlzHistogramMatchObj
\par Name
WlzHistogramMatchObj - Histogram matches a Woolz domain object.
\par Synopsis
\verbatim
WlzHistogramMatchObj  -m<target>
           [-o<output object file>] [-h]
           [-D] [-s<smoothing>] [-t<smoothing>]
           [-i] [-d<min>,<max>] [<input object file>]

\endverbatim
\par Options
<table width="500" border="0">
  <tr>
    <td><b>-D</b></td>
    <td>dither pixel values when mapping.</td>
  </tr>
  <tr>
    <td><b>-m</b></td>
    <td>target histogram object file.</td>
  </tr>
  <tr>
    <td><b>-s</b></td>
    <td>input object histogram smoothing.</td>
  </tr>
  <tr>
    <td><b>-t</b></td>
    <td>target histogram smoothing.</td>
  </tr>
  <tr>
    <td><b>-i</b></td>
    <td>independent planes.</td>
  </tr>
  <tr>
    <td><b>-d</b></td>
    <td> histogram distance range.</td>
  </tr>
  <tr>
    <td><b>-o</b></td>
    <td>Output object file name.</td>
  </tr>
  <tr>
    <td><b>-h</b></td>
    <td>Help - print help message</td>
  </tr>
  <tr>
    <td><b>-v</b></td>
    <td>Verbose operation</td>
  </tr>
</table>
By  default  the  input  object is read from the standard input and the
output object  is written to the standard output.
\par
The histogram distance range limits are comma separated and default  to
0.0,1.0.
\par
If   a  histogram  distance  range  is  given,  then  only  the  domain
objects/planes which have a histogram within that distance  range  from
the target histogram are matched.

\par Description
Modifies  the  grey values of the input object so that the histogram of
       the modified object's grey values matches the target histogram.
\par
       The smoothing  factor  is  the  low-pass  gaussian  convolution  kernel
       half-height  full-width  in  bins  (default 0), used to smooth the his-
       togram(s) before matching.

\par Examples
\verbatim
# An example which uses WlzHistogramMatchObj and WlzHistogramObj to
# match the histogram of one domain object to another.
# an object.

WlzHistogramObj refObj.wlz | \
WlzHistogramMatchObj -m - -o matchedObj.wlz myObj.wlz

# An example which uses WlzHistogramMatchObj to match all planes of
# a 3D image to plane 100.

WlzClipObjToBox -z 100,100 my3Dobj.wlz| \
WlzHistogramObj - | \
WlzHistogramMatchObj -m - -o matched3DObj.wlz my3Dobj.wlz
\endverbatim

\par File
\ref WlzHistogramMatchObj.c "WlzHistogramMatchObj.c"
\par See Also
\ref wlzhistogramobj "WlzHistogramObj(1)",
\ref wlzclipobjtobox "WlzClipObjToBox(1)",
\ref WlzHistogramMatchObj "WlzHistogramMatchObj(3)"
*/

#ifndef DOXYGEN_SHOULD_SKIP_THIS
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <Wlz.h>

extern int      getopt(int argc, char * const *argv, const char *optstring);
 
extern char     *optarg;
extern int      optind,
                opterr,
                optopt;

int             main(int argc, char **argv)
{
  int		idx,
  		option,
		dither = 0,
		smoothingSrcHist = 0,
		smoothingTargetHist = 0,
		independentFlag = 0,
		ok = 1,
		usage = 0;
  double	distVal[2];
  WlzErrorNum	errNum = WLZ_ERR_NONE;
  FILE		*fP = NULL;
  WlzObject	*inObj = NULL,
		*outObj = NULL,
		*matchObj = NULL;
  char 		*outObjFileStr,
		*matchObjFileStr,
  		*inObjFileStr;
  const char	*errMsg;
  char		*distStr[2];
  static char	optList[] = "Dm:s:t:o:id:h",
		outObjFileStrDef[] = "-",
  		inObjFileStrDef[] = "-";

  opterr = 0;
  inObjFileStr = inObjFileStrDef;
  matchObjFileStr = NULL;
  outObjFileStr = outObjFileStrDef;
  distVal[0] = 0.0;
  distVal[1] = 1.0;
  while(ok && ((option = getopt(argc, argv, optList)) != -1))
  {
    switch(option)
    {
      case 'D':
        dither = 1;
	break;
      case 'm':
        matchObjFileStr = optarg;
	break;
      case 's':
        if(sscanf(optarg, "%d", &smoothingSrcHist) != 1)
	{
	  usage = 1;
	  ok = 0;
	}
	break;
      case 't':
        if(sscanf(optarg, "%d", &smoothingTargetHist) != 1)
	{
	  usage = 1;
	  ok = 0;
	}
	break;
      case 'o':
        outObjFileStr = optarg;
	break;
      case 'i':
      	independentFlag = 1;
      case 'd':
        if(optarg)
	{
	  while(*optarg && isspace(*optarg))
	  {
	    ++optarg;
	  }
	  if(*optarg == ',')
	  {
	    distStr[0] = NULL;
	    distStr[1] = strtok(optarg, ",");
	  }
	  else
	  {
	    distStr[0] = strtok(optarg, ",");
	    distStr[1] = strtok(NULL, ",");
	  }
	  if((distStr[0] == NULL) && (distStr[1] == NULL))
	  {
	    usage = 1;
	    ok = 0;
	  }
	  else
	  {
	    idx = 0;
	    while(ok && (idx < 2))
	    {
	      if(distStr[idx] && (sscanf(distStr[idx], "%lg",
	      				 distVal + idx) != 1))
	      {
	        usage = 1;
		ok = 0;
	      }
	      ++idx;
	    }
	  }
	}
	break;
      case 'h':
      default:
        usage = 1;
	ok = 0;
	break;
    }
  }
  if((inObjFileStr == NULL) || (*inObjFileStr == '\0') ||
     (matchObjFileStr == NULL) || (*matchObjFileStr == '\0') ||
     (outObjFileStr == NULL) || (*outObjFileStr == '\0'))
  {
    ok = 0;
    usage = 1;
  }
  if(ok && (optind < argc))
  {
    if((optind + 1) != argc)
    {
      usage = 1;
      ok = 0;
    }
    else
    {
      inObjFileStr = *(argv + optind);
    }
  }
  if(ok)
  {
    errNum = WLZ_ERR_READ_EOF;
    if(((fP = (strcmp(matchObjFileStr, "-")?
        fopen(matchObjFileStr, "r"): stdin)) == NULL) ||
       ((matchObj= WlzAssignObject(WlzReadObj(fP, &errNum), NULL)) == NULL))
    {
      ok = 0;
      (void )WlzStringFromErrorNum(errNum, &errMsg);
      (void )fprintf(stderr,
		    "%s: failed to read histogram object from file %s (%s).\n",
		    *argv, matchObjFileStr, errMsg);
    }
    if(fP && strcmp(matchObjFileStr, "-"))
    {
      fclose(fP);
    }
  }
  if(ok)
  {
    errNum = WLZ_ERR_READ_EOF;
    if((inObjFileStr == NULL) ||
       (*inObjFileStr == '\0') ||
       ((fP = (strcmp(inObjFileStr, "-")?
	      fopen(inObjFileStr, "r"): stdin)) == NULL) ||
       ((inObj= WlzAssignObject(WlzReadObj(fP, &errNum), NULL)) == NULL))
    {
      ok = 0;
      (void )WlzStringFromErrorNum(errNum, &errMsg);
      (void )fprintf(stderr,
		     "%s: failed to read object from file %s (%s).\n",
		     *argv, inObjFileStr, errMsg);
    }
    if(fP && strcmp(inObjFileStr, "-"))
    {
      fclose(fP);
    }
  }
  if(ok)
  {
    if((inObj->type != WLZ_2D_DOMAINOBJ) && (inObj->type != WLZ_3D_DOMAINOBJ))
    {
      ok = 0;
      (void )fprintf(stderr,
		 "%s: input object read from file %s is not a domain object\n",
		     *argv, inObjFileStr);
    }
  }
  if(ok)
  {
    if(matchObj->type != WLZ_HISTOGRAM)
    {
      ok = 0;
      (void )fprintf(stderr,
		     "%s: input object read from file %s is not a histogram\n",
		     *argv, inObjFileStr);
    }
  }
  if(ok)
  {
    if((errNum = WlzHistogramSmooth(matchObj,
    				    smoothingTargetHist)) != WLZ_ERR_NONE)
    {
      ok = 0;
      (void )WlzStringFromErrorNum(errNum, &errMsg);
      (void )fprintf(stderr,
      		     "%s: failed to smooth target histogram (%s).\n",
		     *argv, errMsg);
    }
  }
  if(ok)
  {
    outObj = WlzAssignObject(inObj, NULL);
    if((errNum = WlzHistogramMatchObj(outObj, matchObj,
    				      independentFlag, smoothingSrcHist,
    				      distVal[0], distVal[1],
				      dither)) != WLZ_ERR_NONE)
    {
      ok = 0;
      (void )WlzStringFromErrorNum(errNum, &errMsg);
      (void )fprintf(stderr,
      		     "%s: failed to match object to histogram %s).\n",
		     *argv, errMsg);
    }
  }
  if(ok)
  {
    errNum = WLZ_ERR_WRITE_EOF;
    if(((fP = (strcmp(outObjFileStr, "-")?  fopen(outObjFileStr, "w"):
	      				    stdout)) == NULL) ||
       ((errNum = WlzWriteObj(fP, outObj)) != WLZ_ERR_NONE))
    {
      ok = 0;
      (void )WlzStringFromErrorNum(errNum, &errMsg);
      (void )fprintf(stderr,
		     "%s: failed to write output object (%s).\n",
		     *argv, errMsg);
    }
    if(fP && strcmp(outObjFileStr, "-"))
    {
      fclose(fP);
    }
  }
  if(inObj)
  {
    WlzFreeObj(inObj);
  }
  if(matchObj)
  {
    WlzFreeObj(matchObj);
  }
  if(outObj)
  {
    WlzFreeObj(outObj);
  }
  if(usage)
  {
    (void )fprintf(stderr,
    "Usage: %s%sExample: %s%s",
    *argv,
    " -m <histogram object> [-h] [-o<output object>]\n"
    "                              [-D] [-i] [-s#] [-t#] [-d#,#]\n"
    "                              [<input object>]\n"
    "Options:\n"
    "  -D  Dither pixel values when mapping.\n"
    "  -m  Histogram object to match domain objects values to.\n"
    "  -s  Input object histogram smoothing.\n"
    "  -t  Target histogram smoothing.\n"
    "  -o  Output object/data file name.\n"
    "  -i  If the domain object is a 3D objects then the planes are matched\n"
    "      independently of each other.\n"
    "  -d  Histogram distance range for matching.\n"
    "  -h  Help, prints this usage message.\n"
    "Modifies the grey values of the Woolz domain object so that the\n"
    "histogram of the modified object's grey values matches the given\n"
    "histogram object.\n"
    "If a histogram distance range is given then only domain objects/planes\n"
    "within the range are matched. The range is specified as <min>,<max>\n"
    "where both the minimum and maximum are optional. A valid histogram\n"
    "range is contained by the closed interval [0.0-1.0].\n"
    "Smoothing parameters are gaussian convolution half height full widths\n"
    "specified in histogram bins.\n"
    "Objects/data are read from stdin and written to stdout unless the\n"
    "filenames are given.\n",
    *argv,
    " -m myhist.wlz -i -o matched.wlz mydom.wlz\n"
    "The input Woolz domain object is read from mydom.wlz, matched\n"
    "to the histogram read from the file myhist.wlz and written to the\n"
    "file matched.wlz. If the domain object is a 3D object then it's\n"
    "planes are matched independently.\n");
  }
  return(!ok);
}
#endif /* DOXYGEN_SHOULD_SKIP_THIS */
